### PR TITLE
Add goreleaser

### PR DESCRIPTION
**Pull request by josvazg**
Add **goreleaser** to a release GitHub action and disable versioning and release code from runway CI.

Tested on a separate repo to check out how releases would look like in the GitHub releases page.

### CODE REVIEW

**Comment by josvazg** [comment]
Keeps failing, not sure I can reproduce locally. Debugging...

**Comment by josvazg** [comment]
@tompizmor any clues on how can I void this stuttering?
~~~
binary=/home/runner/work/asset-relocation-tool-for-kubernetes/asset-relocation-tool-for-kubernetes/dist/asset-relocation-tool-for-kubernetes_darwin_arm64/relok8s
~~~

I'd rather have something like this:
~~~
binary=/home/runner/work/asset-relocation-tool-for-kubernetes/asset-relocation-tool-for-kubernetes/dist/relok8s_darwin_arm64/relok8s
~~~

Maybe I should fork this on my own repo to see how releases would look like in github before we merge in main. To avoid suprises.

File: .github/workflows/golangci-lint.yml
Comment (petewall) [inline .github/workflows/golangci-lint.yml:40]: Wouldn't this cut a release on every pull request?

Comment (josvazg) [reply .github/workflows/golangci-lint.yml:40]: Maybe yes.

I should probably for the repo and play with this on my own repo to see how it behaves there, how the releases made look like, etc.

**Comment by josvazg** [comment]
BTW I branched this on my own private repo and I am playing with generating releases there, see how they look, if they autocreate tags, etc

File: .github/workflows/golangci-lint.yml
Comment (migmartri) [reply .github/workflows/golangci-lint.yml:40]: You might need to remove the push.branches=main and leave only the tags. See this one https://github.com/bitnami-labs/charts-syncer/blob/master/.github/workflows/release.yaml

File: .github/workflows/golangci-lint.yml
Comment (migmartri) [inline .github/workflows/golangci-lint.yml:22]: This new functionality should probably be it's own workflow (release.yaml) that only runs when you push a tag not attached to the current one (linter), which as @petewall mentioned, it runs in every push to main and pull request.

**Review by migmartri — CHANGES_REQUESTED** [review]

File: .github/workflows/golangci-lint.yml
Comment (josvazg) [reply .github/workflows/golangci-lint.yml:22]: I see

BTW in my tests I saw this release running in all commits, which we do not want. But it is true that only when you tag main and push that tag is when you get a release done.

will try to make those chages happen, but will test them in my private repo first.

Comment (josvazg) [reply .github/workflows/golangci-lint.yml:22]: I have done this in my repo test, seems to work fine.

**Comment by josvazg** [comment]
Now I am blocked on how to make the version setting automationwork well with tags. I was thinking we could remove the `version` file and list the last tag from `git tag`, but it does not seem so easy...

**Comment by josvazg** [comment]
Sample releases on my repo tests:
Screenshot from 2021-12-15 20-05-20
Screenshot from 2021-12-15 20-06-29

**Review by petewall — APPROVED** [review]

File: .github/workflows/golangci-lint.yml
Comment (josvazg) [reply .github/workflows/golangci-lint.yml]: The `release.yaml` is now separated in this PR.

File: .github/workflows/golangci-lint.yml
Comment (josvazg) [reply .github/workflows/golangci-lint.yml]: done

File: .goreleaser.yml
Comment (tompizmor) [inline .goreleaser.yml:1]: I would delete this comment

**Comment by tompizmor** [comment]
> Now I am blocked on how to make the version setting automation work well with tags. I was thinking we could remove the version file and list the last tag from git tag, but it does not seem so easy...

Can't you use ldflags to set the version of the variable as it is done in [charts-syncer](https://github.com/bitnami-labs/charts-syncer/blob/master/.goreleaser.yml#L12)?

**Comment by tompizmor** [comment]
> @tompizmor any clues on how can I void this stuttering?
> 
> ```
> binary=/home/runner/work/asset-relocation-tool-for-kubernetes/asset-relocation-tool-for-kubernetes/dist/asset-relocation-tool-for-kubernetes_darwin_arm64/relok8s
> ```
> 
> I'd rather have something like this:
> 
> ```
> binary=/home/runner/work/asset-relocation-tool-for-kubernetes/asset-relocation-tool-for-kubernetes/dist/relok8s_darwin_arm64/relok8s
> ```
> 
> Maybe I should fork this on my own repo to see how releases would look like in github before we merge in main. To avoid suprises.

did you solve this already? actually I don't know where those paths are defined so I would need to check the documentation

File: .goreleaser.yml
Comment (josvazg) [reply .goreleaser.yml:1]: let me do it inline...

Comment (josvazg) [reply .goreleaser.yml:1]:
```suggestion
```

**Comment by josvazg** [comment]
> > Now I am blocked on how to make the version setting automation work well with tags. I was thinking we could remove the version file and list the last tag from git tag, but it does not seem so easy...
> 
> Can't you use ldflags to set the version of the variable as it is done in [charts-syncer](https://github.com/bitnami-labs/charts-syncer/blob/master/.goreleaser.yml#L12)?

Yes, relok8s was using that already for setting the version inside the binary. The problem was that we were using a version file to track the version and fill ldflags, while goreleaser uses the tag. I tried to get rid of the version file, but it was problematic as I could not get a reliable last tag from git.
In the end I decided to keep the version file, so the version bumper script does the following:

1. Read the current version file to get the previous version.
2. Compile and run next-semver to:
  - Bump the version automatically, if the script was passed no arguments.
  - Or pass the user desired new version to use. The next-semver will complain if the desired version is not SemVer compliant or is considered before the previous version.
3. Update the version file with the new version.
4. Commit the version bump.
5. Git tag and push the new version as "v${version}".

I think this should work for us to have manual decided releases.

**Comment by josvazg** [comment]
BTW I noticed that there is something wrong with the ldflags trick here, I am passing the `VERSION` env var to **goreleaser** and this is supposedly using make to build the images, so that `VERSION` should be used to set the binary version with the ldflags, but it is NOT doing it for some reason, I still get the dev version.

And this is a PITA to test, as I need to do it on a separate private repo, to avoid polluting our releases here.

**Comment by josvazg** [comment]
> 
> did you solve this already? actually I don't know where those paths are defined so I would need to check the documentation

Looking at the final release, those paths do not really matter. Only the files are listed, like in the images I pasted. So I think I will not try to fix any of that.

File: bump-version.sh
Comment (migmartri) [inline bump-version.sh:1]: I'd move away from automating the version bumping. I think it's ok if we manually tag whenever we decide to release. Similar to https://github.com/kubeapps/kubeapps/blob/master/docs/developer/release-process.md#2---create-a-git-tag

File: next-semver/main.go
Comment (migmartri) [inline next-semver/main.go:1]: There is already tooling for this, I'd not build our own i.e https://github.com/fsaintjacques/semver-tool

File: version
Comment (migmartri) [inline version:1]: What's this file used for? I'd like to move away from adding an additional commit with a file each time. What value is this file providing us?

**Review by migmartri — COMMENTED** [review]
Before (or at the same time) enabling this we will need to disable the current release mechanism.

Could you elaborate how the migration process will look like?

Thanks!

**Comment by josvazg** [comment]
> Before (or at the same time) enabling this we will need to disable the current release mechanism.

Yes, I had not figured out yet how to disable the internal CI version auto bumping. The idea was to include it int this same change, that is why this was still a draft. I might ping Pete about that.

> Could you elaborate how the migration process will look like?

My idea was this change **only** removes internal CI version auto bumping. We would replace the rest of the CI functionality with GitHub Actions in a separate change or set of changes.

File: version
Comment (josvazg) [reply version:1]: I could not figure out a way to automate version bumping without registering the current version somehow.

I can try to remove if we remove all bumping helpers.

File: bump-version.sh
Comment (josvazg) [reply bump-version.sh:1]: The good part of that is we remove a lot of code from this change and the repo in general.

The cost is that we could produce bad version tags, that are not semver or not monotonically increasing. Ideally, we should detect those to avoid releasing bad named versions or versions "from the past".

**Comment by josvazg** [comment]
Going to remove the version bumping helpers and the version file recording.

Will need to test this on the separate repo probably.

**Comment by josvazg** [comment]
@petewall I tried to remove the CI doe that handles and auto-bumps the version in commit 78c50a1
I most likely butchered the whole thing, so any help with that part would be appreciated.

File: .gitignore
Comment (migmartri) [inline .gitignore:11]: add a comment indicating that this is a directory of releases 

```
# Releases

```

File: bump-version.sh
Comment (migmartri) [reply bump-version.sh]: Yeah, I guess we need to trust the operator that it's tagging a proper, valid version. We could also create the release a draft so it can be manually checked before releasing.

In any case, I kind of like how Kubeapps suggests to use some tooling to come up with a valid version, it's better than nothing. https://github.com/kubeapps/kubeapps/blob/master/docs/developer/release-process.md#2---create-a-git-tag

in following patches we should document our release process since now it has a manual component.

File: .goreleaser.yml
Comment (migmartri) [inline .goreleaser.yml:18]: this replacements to add capitalization does not feel right, in fact I do not like the filename resulting of it which uses underscore + Capitalized at the same time.

To me the only two replacements here should be the ones related to the arch

**Review by migmartri — APPROVED** [review]
Accepted although ptal at removing windows support and some other pieces of feedback. Thanks!

File: .goreleaser.yml
Comment (josvazg) [reply .goreleaser.yml:13]: I can remove

File: .gitignore
Comment (josvazg) [reply .gitignore:11]: done

File: .goreleaser.yml
Comment (josvazg) [reply .goreleaser.yml:3]: done

File: .goreleaser.yml
Comment (josvazg) [reply .goreleaser.yml:13]: Done

File: .goreleaser.yml
Comment (josvazg) [reply .goreleaser.yml:18]: done

File: .goreleaser.yml
Comment (migmartri) [inline .goreleaser.yml:9]: i believe this option `goos` now is not required since it's the default one.

File: ci/pipeline.yaml
Comment (migmartri) [inline ci/pipeline.yaml:150]: Shouldn't the release also be removed now? Otherwise we are going to release twice?

**Review by migmartri — COMMENTED** [review]
From a quick check on the current pipeline and taking into account what in my opinion is the end goal (migrate to public CI), in this patch I think we are missing

- Update build job to just build the docker image. In this case it might not make sense to build binaries since it has been delegated to goreleaser.
- Update build job to stop writing the version file and releasing to github
image

I think we also need to revisit that last relocate-chart job since it's weird that it's not part of the external tests that we already have.

**Comment by petewall** [comment]
I'll take a peek at the concourse pipeline. What we want to do going forward is:
* On every commit and on every PR:
    * run the source test suite

* On every new docker image from the registry:
    * run the `relocate-chart` job

The reason I like keeping the `relocate-chart` job is not because it's adding anything to the other test suite, but because it's an easy demo example of using relok8s in a concourse pipeline.

**Comment by josvazg** [comment]
@migmartri yep that sounds reasonable, I will try to figure out the mapping between the pipeline graph and the yaml to make that happen.

@petewall yes, the idea for me is to remove in this PR anything from the CI code that we have moved to goreleaser & actions: build binaries and release, just as Miguel was mentioning.

**Comment by josvazg** [comment]
BTW I am seeing a new transient issue in tests:
~~~go
--- FAIL: TestGroupChangesByChart (0.00s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0
~~~
It happens sometimes, maybe it is a race condition?

**Comment by josvazg** [comment]
Just noticed that @petewall did the CI changes already.

**Comment by josvazg** [comment]
I added docs and fixed the latest cmment by @migmartri . As @petewall has fixed the CI pipeline, we can now mark this as ready to review.

File: ci/pipeline.yaml
Comment (josvazg) [reply ci/pipeline.yaml]: Done by @petewall

File: .goreleaser.yml
Comment (josvazg) [reply .goreleaser.yml]: Done

**Review by jotadrilo — APPROVED** [review]

**Pull request merged**
